### PR TITLE
fix: MSVC issue

### PR DESCRIPTION
**Pull request by henryiii**
Attempt to fix issue seen in #2821. Pulled out the non-changelog part of that PR too (adding pytest-timeout, corrected mypy config).

CC @YannickJadoul @rwgk @bstaletic

### CODE REVIEW

**Comment by YannickJadoul** [comment]
Cross-linking to #2754

**Review by YannickJadoul — APPROVED** [review]

File: include/pybind11/iostream.h [hunk 1]
```diff
@@ -52,9 +52,12 @@ class pythonbuf : public std::streambuf {
 
                 pywrite(line);
                 pyflush();
+
+                // Placed inside gil_scoped_aquire as a makeshift mutex
```
Comment (bstaletic) [inline include/pybind11/iostream.h:56]: Is "makeshift" the right word? Doesn't it just use the OS primitives, somewhere in the depths of CPython?

**Review by bstaletic — APPROVED** [review]

File: include/pybind11/iostream.h (hunk 1 above)
Comment (henryiii) [reply include/pybind11/iostream.h:56]: I almost dropped it... Probably should have. If it builds, I'll drop it and we give it a second chance to fail. ;)

**Comment by rwgk** [comment]
I didn't follow the conversations enough to know if this PR was expected to also fix the TSAN error reported under #2754, but I just tried, and I still see this with current `master`:
```
  WARNING: ThreadSanitizer: data race llvm/llvm-project/libcxx/include/streambuf:462:13 in std::__tsan::basic_streambuf<char, std::__tsan::char_traits<char> >::xsputn(char const*, long) (pybind11/tests/test_iostream)
  WARNING: ThreadSanitizer: data race llvm/llvm-project/libcxx/include/__string:388:56 in std::__tsan::char_traits<char>::copy(char*, char const*, unsigned long) (pybind11/tests/test_iostream)
```

**Comment by henryiii** [comment]
I saw the output issue again, but at least the hangup stopped, and the issue is rarer than it used to be. Issue was in one job out of 5-6 sets of builds.

**Comment by YannickJadoul** [comment]
Hmmm, so we're still missing something? :-(

**Pull request merged**